### ISSUE TITLE
Reduce code duplication

### ISSUE THREAD

**Issue by Venemo** [sailfishos/sailfish-secrets#35]
I'm just writing it here in order not to forget.  
I noticed there are several places where code is duplicated. So here are my current thoughts, I would welcome some feedback from @blammit and @chriadam on them. We would need to fix these eventually if we don't want it to become unmaintainable.

* `evp_p.h` (will be fixed by Bea's PR)
* plugin loading mechanism is needlessly duplicated between secrets and crypto
* some basic methods and enums are present in both secrets and crypto under different namespaces
* in the masterlock branch the plugin locking is duplicated between different plugin types

Possible solutions:

1. Make either crypto depend on secrets, or secrets on crypto. Put all basic declarations in one, and use those in the other.
2. Create a common base for both secrets and crypto. Put all basic declarations there.
3. Reduce the functional overlapping between the two somehow.

What do you guys think?

**Comment by blammit** [comment]
Yes it would be ideal to rework things to greatly reduce the duplication. My understanding is that secrets is about storage and therefore it would depend on crypto rather than the other way around; so, is it a problem if secrets has a hard dependency on crypto? If not, then take that path, and otherwise make a common library that both can use. @chriadam would have a clearer overview of this.

**Comment by chriadam** [comment]
I do agree that the various plugin types should probably extend from a common base class where possible, and that the openssl code duplication should be minimised (the _common.cpp file already does this to a large extent, but there is more to do, e.g. the evp.h as you pointed out already).

In regards to the other points raised:

1) I would greatly prefer to avoid any build time requirement between Secrets and Crypto client API libraries.  There are many use-cases for Secrets which don't require Crypto, and vice-versa.

2) Common base for "basic declarations" would probably require either polluting Sailfish namespace, or the global namespace, either of which seem like suboptimal options, especially since the type-names (or enum names) which might be shared are likely candidates for collisions in the future.  First step might be to see what basic declarations are in fact used across both - I suspect it's a very small number of things.

3) Some functional overlap exists (specifically, Keys can be stored securely, and as an implementation detail these are stored as Secrets), but cannot really be reduced.

**Comment by chriadam** [comment]
While it is definitely true that to be maximally secure, a plugin will most likely want to implement both EncryptedStorage and Crypto plugin interfaces, that is unrelated to the client application requirements.  

That is, a client which uses ONLY crypto functionality (e.g. store some key via Crypto API, use that key to encrypt/decrypt via Crypto API) should NOT need to link against Secrets (e.g., that would have adverse effects on application start time, due to symbol loading etc, and increase dependency complexity).

Similarly, a client which uses ONLY secrets storage functionality (e.g. to store some username or password securely, via the Secrets API) should NOT need to link against Crypto (again, if that was a requirement, we would be enforcing worse application startup time, and dependency complexity).

These are two separate domains, and it is an implementation detail of the daemon-side implementation of the Crypto domain that there is any crossover between these domains at all.

**Comment by chriadam** [comment]
Let me expand on this: some of our customers (B2B and B2G) may have specific requirements for any applications which use Crypto-related APIs (and in the context of automated tooling which might do verification, linking against a Crypto API or pulling in its symbols is probably tantamount to using that Crypto API, for obvious reasons).  Such applications might have more extensive or more lengthy verification cycles, or be subject to different restrictions (export or IP or otherwise legal).  If an application merely wants to securely store some username/password/email-address/etc data, or software license codes, etc, this sort of separation may be extremely valuable (to avoid costly certification etc).

Let me re-iterate: if all you want to gain is some very small de-duplication, but that comes at the cost of worse application startup performance, increased complexity for client applications along with costlier verification cycles, then my opinion is "No, that is not a trade-off we should make in this case."
